### DISCUSSION TITLE
defineProps with a TS type and optional boolean has inconsistent behavior

### DISCUSSION THREAD

**Discussion by TheDutchCoder** [vuejs/language-tools#5030] [Q&A]
We use TS to define props this way:
```ts
const props = defineProps<{
  foo?: boolean
}>()
```

This all works, but Vue compiles optional booleans as `false`, not as `undefined`.
What's interesting is that in the template this works if you access `props.foo`, but not `foo`:
```html
<template>
  <!-- Throws error as foo can be undefined -->
  <SomeComponent :is-visible="foo" />
</template>

<template>
  <!-- This works as expected -->
  <SomeComponent :is-visible="props.foo" />
</template>
```

Our question is: is this a Vue bug, or a language tools bug?

**Comment by KazariEX** [comment]
This is generally expected, as we should not make this prop required when using the component, in which case please use `props.foo`.

**Reply by TheDutchCoder** [reply]
But Vue defaults it to false, instead of undefined, which is the cause of the inconsistency.

As far as I can tell it compiles it to `{ foo: { type: Boolean, default: false } }` and then the types don't match anymore. 

There's also a lint rule to prevent setting default true explicitly, so that makes it even more inconvenient.

Using `props.` in templates is also not standard behavior.

**Reply by KazariEX** [reply]
Please track vuejs/core#12509.

**Reply by Dylancyclone** [reply]
> Please track [vuejs/core#12509](https://github.com/vuejs/core/pull/12509).

I noticed this PR was recently closed without merging the fix. I'm also encountering this issue, is there a reason why it was closed before it was merged? Is there an alternate fix?

**Reply by KazariEX** [reply]
Boolean casting is a flawed design, I don't want to further complicate the code for it.

**Reply by TheDutchCoder** [reply]
> Boolean casting is a flawed design, I don't want to further complicate the code for it.

Can you please elaborate on this?
This is straight from the Vue docs: "The Boolean absent props will be cast to false. You can change this by setting a default for it — i.e.: default: undefined to behave as a non-Boolean prop."

The reasoning here is obviously that a boolean is always ether true or false, and undefined/null/0/etc will all be considered falsy, which makes sense. Using types, it really only ever can be `undefined`, the others would throw a type error.

If you think this should change, I assume you can discuss that with the Vue core team, but since the Vue compiler has this behaviour, so should the language tools.

**Comment by TheDutchCoder** [comment]
Awesome, thanks!
